Docs: Plugin E2E - Add annotations article (#798)
grafana/plugin-tools: docusaurus/docs/e2e-test-a-plugin/test-a-data-source-plugin/annotation-queries.md

---
id: annotation-queries
title: Annotation queries

> description: Test a custom annotation editor and the execution of annotation queries

Reviewer: ```suggestion
description: Test a custom annotation editor and the execution of annotation queries.
```

keywords:
  - grafana
  - plugins
  - plugin
  - testing
  - e2e
  - data-source
  - annotation queries
sidebar_position: 50
---

## Introduction

> Annotations can be used to mark points on a visualization with events such as `AB test started` and `Marketing campaign started`. If a data source plugin supports annotations, it means that the data source can be used to query for annotation events. Optionally, data sources plugins can implement a custom annotation editor that can be used to craft the annotation query. In many cases, the execution of annotation queries requires different handling than normal data queries, and in those cases you may want to write end-to-end tests that verifies that data source annotation support work as expected.

Reviewer: Re: " In many cases, the execution of annotation queries requires different handling than normal data queries..." - Do we need to say anything about those cases which don't require different handling/testing?

> Annotations can be used to mark points on a visualization with events such as `AB test started` and `Marketing campaign started`. If a data source plugin supports annotations, it means that the data source can be used to query for annotation events. Optionally, data sources plugins can implement a custom annotation editor that can be used to craft the annotation query. In many cases, the execution of annotation queries requires different handling than normal data queries, and in those cases you may want to write end-to-end tests that verifies that data source annotation support work as expected.

Author: basically they don't need to write tests, so think this is fine ^.

### Testing the annotation editor

> If your data source plugin implements a custom annotation editor, you can write tests that verifies that the editor works as expected. If no custom editor is implemented, the built-in in editor will be used and then you don't need to write tests for that.

Reviewer: ```suggestion
If your data source plugin implements a custom annotation editor, you can write tests that verify that the editor works as expected. If you haven't implemented a custom editor, then the plugin will use the built-in in editor. In that case, you don't need to write tests.
```

### Testing the entire annotation query execution flow

> In the next example, we perform an integration test where we test a plugin's entire annotation query data flow.

Reviewer: ```suggestion
In the next example, we perform an integration test where we test a plugin's entire annotation query data flow:
```

```ts title="annotations.spec.ts"
test('should run successfully and display a success alert box when query is valid', async ({
  annotationEditPage,
  page,
  selectors,
  readProvisionedDataSource,
}) => {
  const ds = await readProvisionedDataSource({ fileName: 'datasources.yml' });
  await annotationEditPage.datasource.set(ds.name);
  await page.waitForFunction(() => window.monaco);
  await annotationEditPage.getByTestIdOrAriaLabel(selectors.components.CodeEditor.container).click();
  await page.keyboard.insertText(`select time as time, humidity as text
  from dataset
  where $__timeFilter(time) and humidity > 95`);
  await expect(annotationEditPage.runQuery()).toBeOK();
  await expect(annotationEditPage).toHaveAlert('success');
});
```

#### Testing error scenarios

> If an error occurs in the plugin or if the upstream API returns an error, you may want to capture that and return a meaningful error message to the user.

Reviewer: ```suggestion
If an error occurs in the plugin or if the upstream API returns an error, we recommend that you capture that error and return a meaningful error message to the user.
```
